### PR TITLE
[CALCITE-6592] Add test for RelMdPredicates pull up predicate from UNION when it's input predicates include NULL VALUE

### PR DESCRIPTION
**Pull request by NobiGo**
…n it's input include NULL VALUE

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java [hunk 1]
```diff
@@ -2593,6 +2593,28 @@ private void checkPredicates(RelOptCluster cluster, RelOptTable empTable,
 
   }
 
+  @Test void testPullUpPredicatesFromUnionWithValues1() {
```
Comment (caicancai) [inline core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:2596]: add jira link

**Review by mihaibudiu — APPROVED** [review]

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java (hunk 1 above)
Comment (NobiGo) [reply core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:2599]: Added! Thanks for the review.

**Comment by NobiGo** [comment]
Forch-pushed to rerun CI.

**Review by caicancai — APPROVED** [review]
LGTM

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java [hunk 2]
```diff
@@ -2593,6 +2593,31 @@ private void checkPredicates(RelOptCluster cluster, RelOptTable empTable,
...
+   * RelMdPredicates pull up wrong predicate from UNION when it's input include NULL VALUE</a>. */
+  @Test void testPullUpPredicatesFromUnionWithValues1() {
+    final String sql = "select cast(null as integer) as a\n"
+        + "union all\n"
+        + "select 5 as a";
+    final Union rel = (Union) sql(sql).toRel();
+    final RelMetadataQuery mq = rel.getCluster().getMetadataQuery();
+    RelOptPredicateList inputSet = mq.getPulledUpPredicates(rel);
+    ImmutableList<RexNode> pulledUpPredicates = inputSet.pulledUpPredicates;
+    assertThat(pulledUpPredicates, sortsAs("[OR(null, =($0, 5))]"));
```
Comment (zabetak) [inline core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:2607]: The result does not look 100% accurate. Shouldn't this rather be `OR(IS NULL($0), =($0, 5))`. What happens if there were more columns involved?

**Review by zabetak — COMMENTED** [review]
I have the impression that the problem is actually on the way that we extract predicates from [Values](https://github.com/apache/calcite/blob/8d3cb822fd6180b4d6a338b439f647f0c734e55c/core/src/main/java/org/apache/calcite/rel/metadata/RelMdPredicates.java#L590) and not so on what happens in `Union`. I think we should use the `eqConstant` method instead of plain equals.

**Comment by zabetak** [comment]
If the problem is indeed in `Values` please add specific tests for this expression and update the JIRA/PR accordingly.

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java (hunk 2 above)
Comment (NobiGo) [reply core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:2607]: @zabetak Hi, testPullUpPredicatesFromUnionWithValues3 added for more columns. Maybe we should extract nothing when we can't get a specific value like OR(IS NULL($0), =($0, 5)). WDYT?

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java [hunk 3]
```diff
@@ -2593,6 +2593,79 @@ private void checkPredicates(RelOptCluster cluster, RelOptTable empTable,
...
+    assertThat(pulledUpPredicates, sortsAs("[=($0, 1), =($1, 2), =($2, 3)]"));
+  }
+
+  @Test void testPullUpPredicatesFromValues2() {
+    final String sql = "values(null)";
+    final Values values = (Values) sql(sql).toRel();
+    final RelMetadataQuery mq = values.getCluster().getMetadataQuery();
+    RelOptPredicateList inputSet = mq.getPulledUpPredicates(values);
+    ImmutableList<RexNode> pulledUpPredicates = inputSet.pulledUpPredicates;
+    assertThat(pulledUpPredicates, sortsAs("[=($0, null)]"));
```
[the remark this answers is not in the document]
Comment (NobiGo) [reply core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:2647]: Yes. We can use IS NULL($0) to replace =($0, null). I will take care of it.

[the remark this answers is not in the document]
Comment (NobiGo) [reply core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:2647]: CALCITE-1599 has covered the [IS NULL($0)] replace [=$($0, null)]. IS NULL($0) can eliminate ambiguity caused by [=$($0, null)].

**Comment by zabetak** [comment]
I see that you are performing the changes that we discuss here in another pull request (#3981). This is a bit confusing and I am losing track on where I should comment on.

**Comment by NobiGo** [comment]
> I see that you are performing the changes that we discuss here in another pull request (#3981). This is a bit confusing and I am losing track on where I should comment on.

These are two different questions. So I open another PR for VALUES. But we can talk about it together here.
These are some of my opinions：

For Values:
values(1, 2, 3, null), (1, 2, null, null), (5, 2, 3, null)
We pull up the predicate is [=($1, 2), IS NULL($3)]. 
Do we need to extract [OR(IS NULL($2), =($2, 3))]? I think we don't need it. Because when VALUES size is large, It will be complicated.

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mihaibudiu) [reply core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:2647]: Consider is_not_distinct_from

[the remark this answers is not in the document]
Comment (NobiGo) [reply core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:2647]: @zabetak @mihaibudiu I noticed we have a method to handle this:
1) If field `i` is equal to a constant expression, then we use `EQUALS`
2) If field `i`  is nullable, then we use `IS NOT DISTINCT FROM`
3) If filed `i` is the literal null, then we use `IS NULL`

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdPredicates.java [hunk 4]
```diff
@@ -475,7 +475,7 @@ public RelOptPredicateList getPredicates(Union union, RelMetadataQuery mq) {
     RexNode disjunctivePredicate =
         new RexSimplify(rexBuilder, RelOptPredicateList.EMPTY, executor)
             .simplifyUnknownAs(rexBuilder.makeCall(SqlStdOperatorTable.OR, finalResidualPredicates),
-                RexUnknownAs.FALSE);
```
[the remark this answers is not in the document]
Comment (NobiGo) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdPredicates.java:478]: @zabetak I will check this.

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (zabetak) [reply core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:2647]: Yes, we have a method. See the link on my previous comment: https://github.com/apache/calcite/pull/3975#pullrequestreview-2327527960

[the remark this answers is not in the document]
Comment (NobiGo) [reply core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:2647]: Yes. I missed it.

[the remark this answers is not in the document]
Comment (mihaibudiu) [reply core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:2647]: I think that is not distinct from can be used in all cases

[the remark this answers is not in the document]
Comment (zabetak) [reply core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:2647]: I think the distinction among `EQUALS`, `IS NOT DISTINCT FROM`, and `IS NULL` is not only for correctness but also for precision.

**Comment by zabetak** [comment]
Hey @NobiGo  are you still planning to move this forward?

**Comment by NobiGo** [comment]
> Hey @NobiGo are you still planning to move this forward?

Yes. If I have free time, I will handle it. After fixing the CALCITE-6599(RelMdPredicates should pull up more predicates from VALUES when there are several literals) this PR becomes simple. Because we extract the `IS NULL`  not the `[=($0, null)] ` from the VALUES.

**Comment by mihaibudiu** [comment]
I think that pulling up predicates breaks ASOF joins. The spec requires these joins to have predicates of a very specific shape, where each predicate compares a column on one side with a column on the other side. Pulling up predicates may break this invariant. If there is a quick fix that would prevent ASOF joins from participating in the optimization, maybe it can be included in this PR. Otherwise we need to file a new PR. I think that attempting to optimize some of the existing ASOF examples in asof.iq will exhibit the error.

**Comment by NobiGo** [comment]
@zabetak As the unit test, We always extract the predicate `IS NULL($1)` when `null` or  `comm = null` so the `RexUnknownAs` will not affect the results.
And I have discovered another logic that can be optimized:
`select comm = 2 from emp where comm = 2` can pull up `=($0, true)`
`select comm = 3 from emp where comm = 2` can pull up `=($0, false)`
`select comm = 2 from emp where comm = 2 and empno = 1` can pull up `=($0, true)`
`select comm = 2  empno = 1 and from emp where comm = 2 and empno = 1` can pull up `=($0, true)`
Now the pull up predicate is nothing. WDYT?

**Pull request merged**